valueset.go

```diff
@@ -0,0 +1,326 @@
+	if oldNodeset.ID == "" {
+		root := NodesetData{
+			ID:         newNodeset.ID,
+			Properties: newNodeset.Properties,
+			Labels:     newNodeset.Labels,
+		}
+		insertions = append(insertions, root)
+		rootOp = "insertRoot"
+	} else {
+		// delete DB nodeset if newNodeset is empty
+		if newNodeset.ID == "" {
+			deletions = append(deletions, oldNodeset.ID)
+			rootOp = "deleteRoot"
+		} else if oldNodeset != newNodeset {
+			// update to use newNodeset if oldNodeset is not equal
+			root := NodesetData{
+				ID:         newNodeset.ID,
+				Properties: newNodeset.Properties,
+				Labels:     newNodeset.Labels,
+			}
+			updates = append(updates, root)
+			rootOp = "updateRoot"
+		}
+	}
+	for oid, oldNsData := range oldNodeset.Data {
+		newNsData, exists := newNodeset.Data[oid]
+		if exists {
+			update := NodesetData{
+				ID:         oid,
+				Labels:     make([]string, 0),
+				Properties: make(map[string]interface{}),
+			}
+			changed := false
+			if !lpg.NewStringSet(oldNsData.Labels...).IsEqual(lpg.NewStringSet(newNsData.Labels...)) {
+				// update - setting labels to those in newNodeset
+				update.Labels = newNsData.Labels
+				changed = true
+			}
+			// if props are not equal, update to use newNodeset properties
+			if !reflect.DeepEqual(oldNsData.Properties, newNsData.Properties) {
+				for k, v := range newNsData.Properties {
+					update.Properties[k] = v
+				}
+				changed = true
+			}
+			if changed {
+				updates = append(updates, update)
+			}
+		} else {
+			deletions = append(deletions, oid)
+		}
+	}
+	for nid, newNsData := range newNodeset.Data {
+		_, exists := oldNodeset.Data[nid]
+		if !exists {
+			insertions = append(insertions, newNsData)
+		}
+	}
+	return rootOp, insertions, deletions, updates
+}
+
+func LoadNodeset(tx neo4j.Transaction, nodesetId string) (Nodeset, error) {
+	query := "MATCH (root:`NODESET` {nodeset_id: $id})-[:$id]->(m) return root,m"
+	idrec, err := tx.Run(query, map[string]interface{}{"id": nodesetId})
+	if err != nil {
+		return Nodeset{}, err
+	}
+	ns := Nodeset{
+		ID:     nodesetId,
+		Labels: make([]string, 0),
+		Data:   make(map[string]NodesetData),
+	}
+	first := false
+	for idrec.Next() {
+		rec := idrec.Record()
+		if !first {
+			nsCenter := rec.Values[0].(neo4j.Node)
+			ns.Labels = nsCenter.Labels
+			first = true
+		}
+		nsLeaf := rec.Values[1].(neo4j.Node)
+		var nsLeafEntityId string
+		for k := range nsLeaf.Props {
+			if k == ls.EntityIDTerm {
+				nsLeafEntityId = k
+			}
+		}
+		ns.Data[nsLeafEntityId] = NodesetData{
+			ID:         nsLeafEntityId,
+			Labels:     nsLeaf.Labels,
+			Properties: nsLeaf.Props,
+		}
+	}
+	return ns, nil
+}
+
+func Execute(tx neo4j.Transaction, oldNodeset, newNodeset *Nodeset, rootOp string, inserts, updates []NodesetData, deletions []string) error {
+	type groupedNodesetData struct {
+		labels lpg.StringSet
+		data   []NodesetData
+	}
+	groupByLabel := func(nsD []NodesetData) []groupedNodesetData {
```

Comment:
This is not grouping by label. You are simply copying the nodeset data into groupedNodesetData.
Do this:
groupedNodesetData struct {
   LabelExpr string
   Data []NodesetData
}

m: map[string][]NodesetData
for each nsD {
  sort labels of nsD
  Write a label expression using config
  m[labelExpression]=append(m[labelExpression], nsD)
}
m contains the grouped nodeset data, From m, generate goupredNodesetData
}